### PR TITLE
add `AUTO_CD` functionality

### PR DESCRIPTION
**Pull request by gforsyth**
As suggested in #439 

I couldn't find a better place to insert this other than `base_shell.py` -- it might be a bit too hacky, but I'm open to suggestions.

If user enters the name of a directory, e.g.

```
gil@theo ~ $ /usr/bin
```

then Xonsh will first try to run that command.  When it fails and raises
a XonshError, if the env variable `$AUTO_CD=True` then Xonsh will check
to see if the input is a valid directory and that there is no binary in
`$PATH` that matches the input (this last bit is likely redundant).

If all conditions are met, it reruns the "command" but prepends `cd` to it.

### CODE REVIEW

**Comment by scopatz** [comment]
Hi @gforsyth - thanks a million for putting this in!  I would actually put this in the `xonsh.built_ins.run_subproc()` functions, probably somewhere around [line #525](https://github.com/scopatz/xonsh/blob/master/xonsh/built_ins.py#L525). If autocd is true, if there are no other arguments, if the command is a directory and cannot be found in the aliases or via `locate_binary()`, then replace the command with `cd dir` alias. I think that this is less hacky than sticking it in BaseShell.

**Comment by gforsyth** [comment]
Hey @scopatz , happy to help!  I was sure there must be a less hacky way but couldn't figure it out.  Now it's in `run_subproc()` and working well on my machine.

File: xonsh/built_ins.py [hunk 1]
```diff
@@ -523,6 +523,13 @@ def run_subproc(cmds, captured=True):
             stderr = streams['stderr']
         uninew = (ix == last_cmd) and (not captured)
         alias = builtins.aliases.get(cmd[0], None)
+        if (alias is None
+            and builtins.__xonsh_env__.get('AUTO_CD')
+            and os.path.isdir(cmd[0])
+            and locate_binary(cmd[0], cwd=None) is None
+            and len(cmds)==1):
+            cmd.insert(0, 'cd')
+            return run_subproc(cmds, captured=False)
```
Comment (scopatz) [inline xonsh/built_ins.py]: Does this need to be recursive? I think you can just set `alias = builtins.aliases.get('cd', None)` here instead of returning.

**Comment by scopatz** [comment]
Is this ready to go @gforsyth?

**Comment by gforsyth** [comment]
Should be good.  Tests passing, works as expected.

**Comment by scopatz** [comment]
Thanks a ton @gforsyth - awesome feature to have. We might consider changing the default to True :)

**Pull request merged**